ufl/matrix.py

```diff
@@ -0,0 +1,118 @@
+# -*- coding: utf-8 -*-
+"""This module defines the Matrix class."""
+
+# Copyright (C) 2021 India Marsden
+#
+# This file is part of UFL (https://www.fenicsproject.org)
+#
+# SPDX-License-Identifier:    LGPL-3.0-or-later
+
+
+from ufl.log import error
+from ufl.finiteelement import FiniteElementBase
+from ufl.domain import default_domain
+from ufl.form import BaseForm
+from ufl.argument import Argument
+from ufl.functionspace import AbstractFunctionSpace, FunctionSpace
+from ufl.utils.counted import counted_init
+
+
+# --- The Matrix class represents a matrix, an assembled two form ---
+
+class Matrix(BaseForm):
+    """UFL form argument type: Parent Representation of a form coefficient."""
+
+    __slots__ = (
+        "_count",
+        "_ufl_function_spaces",
+        "_repr",
+        "_hash",
+        "_ufl_shape",
+        "_arguments")
+    _globalcount = 0
+
+    def __getnewargs__(self):
+        return (self._ufl_function_spaces[0], self._ufl_function_spaces[1], self._count)
+
+    def __init__(self, row_space, column_space, count=None):
+        BaseForm.__init__(self)
+        counted_init(self, count, Matrix)
+
+        if isinstance(row_space, FiniteElementBase):
+            # For legacy support for .ufl files using cells, we map
+            # the cell to The Default Mesh
+            element = row_space
+            domain = default_domain(element.cell())
+            row_space = FunctionSpace(domain, element)
+        elif not isinstance(row_space, AbstractFunctionSpace):
+            error("Expecting a FunctionSpace or FiniteElement as the row space.")
+
+        if isinstance(column_space, FiniteElementBase):
+            # For legacy support for .ufl files using cells, we map
+            # the cell to The Default Mesh
+            element = column_space
+            domain = default_domain(element.cell())
+            column_space = FunctionSpace(domain, element)
+        if not isinstance(column_space, AbstractFunctionSpace):
+            error("Expecting a FunctionSpace or FiniteElement as the column space.")
+
```

Comment:
Don't bother doing backward compat for new features, just require `AbstractFunctionSpace` objects.